lib/srv/app/common/audit.go

```diff
@@ -0,0 +1,251 @@
+		ConnectionMetadata: apievents.ConnectionMetadata{
+			RemoteAddr: identity.ClientIP,
+		},
+		AppMetadata: apievents.AppMetadata{
+			AppURI:        app.GetURI(),
+			AppPublicAddr: app.GetPublicAddr(),
+			AppName:       app.GetName(),
+		},
+	}
+	return trace.Wrap(a.EmitEvent(ctx, event))
+}
+
+// OnSessionChunk is called when a new session chunk is created.
+func (a *audit) OnSessionChunk(ctx context.Context, sessionCtx *SessionContext, serverID string) error {
+	event := &apievents.AppSessionChunk{
+		Metadata: apievents.Metadata{
+			Type:        events.AppSessionChunkEvent,
+			Code:        events.AppSessionChunkCode,
+			ClusterName: sessionCtx.Identity.RouteToApp.ClusterName,
+		},
+		ServerMetadata: apievents.ServerMetadata{
+			ServerID:        serverID,
+			ServerNamespace: apidefaults.Namespace,
+		},
+		SessionMetadata: apievents.SessionMetadata{
+			SessionID: sessionCtx.Identity.RouteToApp.SessionID,
+			WithMFA:   sessionCtx.Identity.MFAVerified,
+		},
+		UserMetadata: sessionCtx.Identity.GetUserMetadata(),
+		AppMetadata: apievents.AppMetadata{
+			AppURI:        sessionCtx.App.GetURI(),
+			AppPublicAddr: sessionCtx.App.GetPublicAddr(),
+			AppName:       sessionCtx.App.GetName(),
+		},
+		SessionChunkID: sessionCtx.ChunkID,
+	}
+	return trace.Wrap(a.EmitEvent(ctx, event))
+}
+
+// OnRequest is called when an app request is sent during the session and a response is received.
+func (a *audit) OnRequest(ctx context.Context, sessionCtx *SessionContext, req *http.Request, res *http.Response, re *endpoints.ResolvedEndpoint) error {
+	if sessionCtx.App.IsDynamoDB() {
+		return trace.Wrap(a.onDynamoDBRequest(ctx, sessionCtx, req, res, re))
+	}
+	event := &apievents.AppSessionRequest{
+		Metadata: apievents.Metadata{
+			Type: events.AppSessionRequestEvent,
+			Code: events.AppSessionRequestCode,
+		},
+		AppMetadata:        *MakeAppMetadata(sessionCtx.App),
+		Method:             req.Method,
+		Path:               req.URL.Path,
+		RawQuery:           req.URL.RawQuery,
+		StatusCode:         uint32(res.StatusCode),
+		AWSRequestMetadata: *MakeAWSRequestMetadata(req, re),
+	}
+	return trace.Wrap(a.EmitEvent(ctx, event))
+}
+
+// onDynamoDBRequest is called when a DynamoDB app request is sent during the session.
+func (a *audit) onDynamoDBRequest(ctx context.Context, sessionCtx *SessionContext, req *http.Request, res *http.Response, re *endpoints.ResolvedEndpoint) error {
+	// Try to read the body and JSON unmarshal it.
+	// If this fails, we still want to emit the rest of the event info; the request event Body is nullable, so it's ok if body is left nil here.
+	var body *apievents.Struct
+	if jsonBody, err := io.ReadAll(req.Body); err != nil {
+		a.log.WithError(err).Warn("Failed to read DynamoDB request body.")
+	} else {
+		s := &apievents.Struct{}
+		if err := s.UnmarshalJSON(jsonBody); err != nil {
+			a.log.WithError(err).Warn("Failed to decode DynamoDB request JSON body.")
+		} else {
+			body = s
+		}
+	}
```

Comment:
I think you're right it would be better as a function. Since I'm going to implement database-access for dynamodb today, I will need the same functionality there as well. Therefore, I've defined a new function in `lib/utils/aws/aws.go` for extracting the JSON body.

Reply:
@tobiaszheller I know you already approved, but could you take a look at that new function `UnmarshalRequestBody` I added to aws utils? Just want to make sure you approve of that change since it wasn't exactly what you suggested here. I've marked this convo unresolved again for that reason.